embassy-stm32/src/dcmi.rs

```diff
@@ -0,0 +1,465 @@
+
+        unborrow!(
+            peri, dma, irq, d0, d1, d2, d3, d4, d5, d6, d7, d8, d9, d10, d11, d12, d13, v_sync,
+            h_sync, pixclk
+        );
+
+        d0.configure();
+        d1.configure();
+        d2.configure();
+        d3.configure();
+        d4.configure();
+        d5.configure();
+        d6.configure();
+        d7.configure();
+        d8.configure();
+        d9.configure();
+        d10.configure();
+        d11.configure();
+        d12.configure();
+        d13.configure();
+
+        v_sync.configure();
+        h_sync.configure();
+        pixclk.configure();
+
+        let edm = match (
+            d8.pin().is_some(),
+            d9.pin().is_some(),
+            d10.pin().is_some(),
+            d11.pin().is_some(),
+            d12.pin().is_some(),
+            d13.pin().is_some(),
+        ) {
+            (true, true, true, true, true, true) => 0b11, // 14 bits
+            (true, true, true, true, false, false) => 0b10, // 12 bits
+            (true, true, false, false, false, false) => 0b01, // 10 bits
+            (false, false, false, false, false, false) => 0b00, // 8 bits
+            _ => {
+                #[cfg(feature = "defmt")]
+                defmt::panic!("Invalid pin configuration.");
+                #[cfg(not(feature = "defmt"))]
+                core::panic!("Invalid pin configuration.");
+            }
+        };
+
+        unsafe {
+            peri.regs().cr().modify(|r| {
+                r.set_cm(true); // disable continuous mode (snapshot mode)
+                r.set_ess(use_embedded_synchronization);
+                r.set_pckpol(pixclk_polarity == PixelClockPolarity::RisingEdge);
+                r.set_vspol(vsync_level == VSyncDataInvalidLevel::High);
+                r.set_hspol(hsync_level == HSyncDataInvalidLevel::High);
+                r.set_fcrc(0x00); // capture every frame
+                r.set_edm(edm); // extended data mode
+            });
+            peri.regs().ier().modify(|r| {
+                r.set_ovr_ie(true);
+                r.set_frame_ie(true);
+            });
+        }
+
+        unsafe {
+            Self {
+                inner: PeripheralMutex::new_unchecked(irq, &mut state.0, move || StateInner {
+                    phantom: PhantomData,
+                    peri,
+                    dma,
+                    isr_result: None,
+                    waker: WakerRegistration::new(),
+                }),
+            }
+        }
+    }
+
+    unsafe fn toggle(peri: &T, enable: bool) {
+        peri.regs().cr().modify(|r| {
+            r.set_enable(enable);
+            r.set_capture(enable);
+        })
+    }
+
+    pub async fn capture(&mut self, buffer: &mut [u32]) -> Result<(), Error> {
+        self.inner.with(|state| {
+            let channel = &mut state.dma;
+
+            let r = state.peri.regs();
+            let src = r.dr().ptr() as *mut u32;
+            unsafe { channel.start_read(channel.request(), src, buffer) };
```

Comment:
Here you're starting the DMA, but never stop it. The user can start a capture and then cancel the operation (by dropping the future for this `async fn` before it completes). This releases the borrow on `buffer`, so the user can then deallocate it. However DMA has not been stopped, so it'll keep writing to it, overwriting whatever and causing UB.

If you use the safe wrapper `dma::read()` instead, it returns a future that you can await, and that *stops the DMA when dropped*. That'll cause DMA to stop automatically if the user drops the future for the `async fn capture()`, preventing the UB (because dropping a future drops all its contents recursively).